client/HeroMovementController.cpp

```diff
@@ -0,0 +1,378 @@
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+#include "StdInc.h"
+#include "HeroMovementController.h"
+
+#include "CGameInfo.h"
+#include "CMusicHandler.h"
+#include "CPlayerInterface.h"
+#include "PlayerLocalState.h"
+#include "adventureMap/AdventureMapInterface.h"
+#include "eventsSDL/InputHandler.h"
+#include "gui/CGuiHandler.h"
+#include "gui/CursorHandler.h"
+#include "mapView/mapHandler.h"
+
+#include "../CCallback.h"
+
+#include "../lib/pathfinder/CGPathNode.h"
+#include "../lib/mapObjects/CGHeroInstance.h"
+#include "../lib/RoadHandler.h"
+#include "../lib/TerrainHandler.h"
+#include "../lib/NetPacks.h"
+
+bool HeroMovementController::isHeroMovingThroughGarrison(const CGHeroInstance * hero, const CArmedInstance * garrison) const
+{
+	if(!duringMovement)
+		return false;
+
+	if(!LOCPLINT->localState->hasPath(hero))
+		return false;
+
+	if(garrison->visitableAt(LOCPLINT->localState->getPath(hero).lastNode().coord))
+		return false; // hero want to enter garrison, not pass through it
+
+	return true;
+}
+
+bool HeroMovementController::isHeroMoving() const
+{
+	return duringMovement;
+}
+
+void HeroMovementController::onPlayerTurnStarted()
+{
+	assert(duringMovement == false);
+	assert(stoppingMovement == false);
+	duringMovement = false;
+	currentlyMovingHero = nullptr;
+}
+
+void HeroMovementController::onBattleStarted()
+{
+	// when battle starts, game will send battleStart pack *before* movement confirmation
+	// and since network thread wait for battle intro to play, movement confirmation will only happen after intro
+	// leading to several bugs, such as blocked input during intro
+	movementAbortRequested();
+}
+
+void HeroMovementController::showTeleportDialog(const CGHeroInstance * hero, TeleportChannelID channel, TTeleportExitsList exits, bool impassable, QueryID askID)
+{
+	// Player entered teleporter
+	// Check whether hero that has entered teleporter has paths that goes through teleporter and select appropriate exit
+	// othervice, ask server to select one randomly by sending invalid (-1) value as answer
+	assert(waitingForQueryApplyReply == false);
+	waitingForQueryApplyReply = true;
+
+	if(!LOCPLINT->localState->hasPath(hero))
+	{
+		// Hero enters teleporter without specifying exit - select it randomly
+		LOCPLINT->cb->selectionMade(-1, askID);
+		return;
+	}
+
+	const auto & heroPath = LOCPLINT->localState->getPath(hero);
+	const auto & nextNode = heroPath.nextNode();
+
+	for(size_t i = 0; i < exits.size(); ++i)
+	{
+		const auto * teleporter = LOCPLINT->cb->getObj(exits[i].first);
+
+		if(teleporter && teleporter->visitableAt(nextNode.coord))
+		{
+			// Remove this node from path - it will be covered by teleportation
+			//LOCPLINT->localState->removeLastNode(hero);
+			LOCPLINT->cb->selectionMade(i, askID);
+			return;
+		}
+	}
+
+	assert(0); // exit not found? How?
```

Comment:
one-way monolith may not have exit on the map

Reply:
Will recheck, but in that case teleport dialog probably is not created to begin with.